### ISSUE TITLE
Missing .h hides *ALL* errors in .c source (formerly #2403)

### ISSUE THREAD

**Issue by nurbles** [microsoft/vscode-cpptools#2406]
Since #2403 was summarily closed without any attempt at understanding the solution suggested/requested, I'm starting a new thread to explain...

I understood WHY _**MOST**_ errors are hidden/not reported when a missing header is detected.  What I don't understand is why anyone thinks that adding a dark green squiggly under the #include line on a dark gray background is useful.  Especially since that single line MUST actually be in the window for the mark to have any utility at all -- even if it were visible with the default theme (which has the dark gray background.)

I was asking for the _**OTHER**_ two error cues to still be used for the bad #include line:  Mark the source file with an error (**ONE** error for each bad #include would be best) in the project file list, as is done when _any other error_ exists in a file.  Also, add a red marker in the scroll bar showing where the offending #include line is within the file, again as is done for any other error.

Note that I am specifically _**NOT**_ asking for the remainder of the file to be processed and hundreds of errors to show, but I really, REALLY, **REALLY** want to see that there is a "key" error that is preventing other errors from showing (and may also be the reason why other features, such as struct member completion from working.)

I apologize if this wasn't immediately clear in my original post and I hope it is clear now.

**Comment by sean-mcmanus** [comment]
We sometimes close issues that appear resolved to us under the assumption that users will reply to the closed issue with more info if they believe it was closed incorrectly and then we re-open the issue -- this prevents us from having too many open issues that become harder for us to track. So starting a new issue isn't necessary, but okay as well.

I've never had any problem seeing the green squiggles on the dark background or heard complaints about it, but the colors can be customized for users who do: https://code.visualstudio.com/docs/getstarted/theme-color-reference .

Also, the message is shown in the Problems pane at the bottom if that is made visible.

The lack of color in the scroll bar appears to be "by design" in VS Code for messages classified with "Information". VS Code shows the files in red text if they have errors, so if include paths aren't configured, we probably don't want to show all the opened files as red by default.

Was there a problem with the solution of setting the intelliSenseEngineFallback to "Disabled"?

We could add some setting to enabling changing the "#include error detected" message classification category between Info/Warning/Error. Maybe we could change the default level to "Warning" -- I believe we originally classified it as Info because warnings squiggles were red originally, but they're green now.

**Comment by nurbles** [comment]
1. In other commenting systems "closed" typically means no more comments may be added, so instead I started this thread.

2. From the context, the intelliSenseEngineFallback "Disabled" sounded like it would lead to all possible errors being flagged in the file.  That, combined with my frustration at not being allowed to explain what I was requesting caused me to ignore the setting.  When you asked about it I gave it a shot and it does exactly what I was requesting.  If only that had been clearer in the previous thread's comment.  Thanks for being patient and re-recommending the setting which appears to do what you suggest at the end by changing the #include issue from info to error.

3. I'm still getting up to speed with VSCode but I'm not finding documentation for the settings very helpful ... yet.  I need to completely change the way I think in order to make sense of some of the documentation.  However, I'm immensely frustrated by the apparent impossibility of configuring my workspace such that I can click on a problem to jump to the appropriate line.  The **ONLY** setting that appears in the default workspace when I first opened to edit is folder and path -- neither of which can I find documented _**ANYWHERE**_.  But not matter their setting, clicking a problem yields a message about not being about to open file:///filename.ext and I can assure you that I have **NOT** set the path to "///"!  I've found many people struggling with this problem and a couple Windows-specific solutions that don't help me under Ubuntu, but no linux solutions.  I guess that is another issue to report.

**Comment by sean-mcmanus** [comment]
1. Yeah, GitHub allows comments on closed issues unless the issue is marked as "locked". I'm not sure what other commenting systems there are, but the VS 2017 support page used to not allow comments after closing, which I complained about a year ago...not sure if they fixed that: https://developercommunity.visualstudio.com/content/problem/76224/dont-let-someone-close-an-issue-until-theres-a-pro.html .

3. Documentation on which settings? We publish the docs on the C_Cpp settings and VS Code itself owns the others ones. Clicking on a line in the Problems pane should jump to the appropriate line (if the problem matcher calculated the line correctly) -- can you give more repro details? I don't know what the "folder" and "path" settings are. Are you talking about settings.json or c_cpp_properties.json (or launch.json, tasks.json)? VS Code has a fancy editor in "preview" and the normal json file editor. The file:/// issue is a problem with problem matcher and tool that outputs the problem -- VS Code uses "problem matchers" to auto-generate problems from tools, and certain tools may output incompatible problems for the problem matchers available. Our C/C++ extension just ships a gcc problem matcher that uses a regular expression to extract the line number of the error based on the gcc error output (see https://github.com/Microsoft/vscode-cpptools/blob/master/Extension/package.json ). We have some bugs logged related to problem matchers: https://github.com/Microsoft/vscode-cpptools/issues?utf8=%E2%9C%93&q=is:issue+is:open+problem+matcher . The gcc problem matcher assumes the path is relative, so issue https://github.com/Microsoft/vscode-cpptools/issues/1915 might be relevant to you -- the workaround is to use something like the following in your build task:
```
"problemMatcher": {
    "base": "$gcc",
    "fileLocation": "absolute"
}
```

**Comment by nurbles** [comment]
- Documentation: I was trying to read through documentation for setting colors in vscode and getting more and more lost because I don't "speak the language" of the settings and/or don't yet understand the hierarchy of the settings (vscode vs extension vs user vs etc.)  So my doc problem is hopefully more me than anything else.

- The folder/path settings are what I found when I opened rm23p.code-workspace:
`
    {
       "folders": [
          {
             "path": "./"
          }
       ],
       "settings":
       {
       }
    }
`
I know the workspace isn't an extension issue, it was just an example of something I've been searching for in the documentation that simply does not exist.  There are pages of workspace settings documented, just neither of those two.

- Clicking a problem to jump to the line of code appears to work with the problems generated by the C extension while editing seems to work OK, but like you say, it doesn't work with compiler output.  I am already using
    `"fileLocation": ["absolute"]`
Which was how I found an example of how to use make to build my project.  If vscode with the C extension has a "better" way to define the project (with at least release and debug configurations, plus clean options for both) I'd like to find the documentation.  Instead it just seemed to hand all of the workspace source to gcc and hope for the best.

**Comment by nurbles** [comment]
FWIW, I changed my problem matcher definition to EXACTLY match your example (by cutting your example and pasting it into the task.json file) and still get the "unable to open file:///filename.ext" message for compiler errors.  I'm curious why the C extension problems are listed with the file name showing as "main.c   " while compiler errors have a separate section titled "main.c   /" ?  I'm guessing that is related to the root cause of this issue.

**Comment by nurbles** [comment]
Oh, I forgot to mention that setting the intelliSenseEngineFallback to "Disabled" enabled _**ALL**_ errors to be displayed again instead of just making the bad #include line the only error.  I think just flagging the bad #include as an error would be a better solution in the long run.  That way it would always be visible, even if you don't have the file OPEN (perhaps you renamed or removed an include one of your files requires.)  I hope you can suggest this to the extension developers and pray that they see its usefulness.

**Comment by sean-mcmanus** [comment]
Oh, the folders/settings are the "workspace settings". "folders" is read-only and lists the folders that are part of a multi-root workspace and "settings" is just the settings that apply to all the folders in the workspace.

We plan to add some out-of-the-box C/C++ build tasks as part of https://github.com/Microsoft/vscode-cpptools/issues/2359 .

Yeah, if we added some "includePathErrorSeverity" setting with values "Information" (current default), "Warning" (new default?), and "Error" that would avoid showing all the errors but increase the visibility of the issue. Does that sound sufficient?

FYI, we are the extension developers (and testers/documenters/etc.), so we'll consider this change. The VS Code (and GitHub) teams generally don't have any separate "support" staff like the Visual Studio support page uses.

**Comment by nurbles** [comment]
It is nice to know that the actual developers are looking at these forums.  I'm so accustomed to going through an intermediate layer in Microsoft forums that I just assumed this might be more of the same.

Yes, I like your idea of something like an "includePathErrorSeverity" setting.  It sounds like it would allow the #include error to be elevated enough to flag the issue at the file level so that it would be visible even when the file was closed, as other errors are when I "break" a structure definition in a header, for example.

FWIW, I found a good example of one of my documentation issues:  Trying to find out the "common colors" for the editor.tokenColorCustomizations setting takes me in a nice documentation circle and none of the pages in that circle do more than offer a couple examples of what "common" probably means.  It sounds like there is a specific list of common syntax element/token color names defined, but apparently we must guess them.  (smile)  Again, I suspect this is outside the realm of this extension so it is only here as an example of the type of documentation issue I was encountering.

Thanks for all of your help!  I will do my best to keep quiet for a while.  <smile>
